File: S2/Programmation/TD1/stats.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;
binmode(STDIN, "utf8");
binmode(STDOUT, "utf8");
binmode(STDERR, "utf8");

# where the files are
my $dir = "corpustest" ;

# read only txt
my @files = <$dir/*.txt> ;

my %words;

# for each of the files
foreach my $file(@files) {
  # open file (readmode)
  open(INPUT, '<:utf8', $file);

  my $token_counter;
  # read content
  while (<INPUT>) {
    # get token from file
    # it is already tokenized with spaces, so easy to split
    foreach my $token(split(" ", $_)) {
      # create the hash of words that will autoincrement for same words
      $words{ $token }++ ;
      # increment total number of token
      $token_counter++;
    }
  }
  close(INPUT);

  # ok, so the various results are :
  my $total_tokens = $token_counter;
  my $total_words  = keys(%words);
  my $ratio        = $total_tokens / $total_words ;

  # now we create a new file to store the answers
  # new file name is a little modification of input :
  # ie remove the 3 last char (that is "txt")
  # and add the new extension name "freq"
  $file = substr($file, 0, -3) . "freq" ;

  # open that new file for writing
  open(OUTPUT, '>:utf8', $file);
  # and print the results in (tab separated)
  print OUTPUT $total_words . "\t" . $total_tokens . "\t" . $ratio ."\n";
  close(OUTPUT);
}

File: S2/Programmation/TD1/stats.rb
#!/usr/bin/env ruby

files = Array.new

Dir.glob("corpustest/*.txt") do |file|
  files << file
end

class Stats

  def initialize(file)
    puts file
    input = File.open(file, "r")
    @word_list = Array.new
    input.each do |line|
      words = line.split(" ")
      words.each do |word|
        @word_list << word
      end
    end
    input.close
  end
  
  def token
    $token = @word_list.length
    puts "token: #{$token}"
  end

  def words
    freqs = Hash.new { |h,k| freqs[k] = 0 }
    @word_list.each { |e| freqs[e] += 1 }
    $word_total = freqs.length

    puts "words: #{$word_total}"
  end

  def ratio
    ratio = $word_total.to_f / $token
    puts "ratio: #{ratio}"
  end

end
 

files.each do |file|
  fichier = Stats.new(file)
  fichier.token
  fichier.words
  fichier.ratio
  puts
end

